File: server/agent/langgraph/agents/OrchestratorAgent.ts
import { z } from "zod";
import OpenAI from "openai";
import { BaseAgent, BaseAgentConfig, AgentTask, AgentResult, AgentCapability, AGENT_REGISTRY } from "./types";

const xaiClient = new OpenAI({
  baseURL: "https://api.x.ai/v1",
  apiKey: process.env.XAI_API_KEY,
});

const DEFAULT_MODEL = "grok-4-1-fast-non-reasoning";

export class OrchestratorAgent extends BaseAgent {
  constructor() {
    const config: BaseAgentConfig = {
      name: "OrchestratorAgent",
      description: "Super agent that coordinates and delegates tasks to specialized agents. Analyzes requests, creates execution plans, and orchestrates multi-agent workflows.",
      model: DEFAULT_MODEL,
      temperature: 0.3,
      maxTokens: 4096,
      systemPrompt: `You are the OrchestratorAgent - a super-intelligent coordinator for a multi-agent system.

Your responsibilities:
1. Analyze complex user requests and break them into subtasks
2. Route tasks to the most appropriate specialized agents
3. Coordinate multi-agent workflows and handle dependencies
4. Aggregate results from multiple agents
5. Handle errors and implement retry strategies
6. Optimize execution order for efficiency

Available specialized agents:
- ResearchAgent: Web research, information gathering, fact-checking
- CodeAgent: Code generation, review, refactoring, debugging
- DataAgent: Data analysis, transformation, visualization
- ContentAgent: Content creation, document generation
- CommunicationAgent: Email, notifications, messaging
- BrowserAgent: Autonomous web navigation and interaction
- DocumentAgent: Document processing and manipulation
- QAAgent: Testing, validation, quality assurance
- SecurityAgent: Security audits, encryption, compliance

When delegating tasks, provide clear instructions and context. Monitor progress and handle failures gracefully.`,
      tools: ["plan", "orchestrate", "decide", "reflect"],
      timeout: 300000,
      maxIterations: 50,
    };
    super(config);
  }

  async execute(task: AgentTask): Promise<AgentResult> {
    const startTime = Date.now();
    this.updateState({ status: "running", currentTask: task.description, startedAt: new Date().toISOString() });

    try {
      const plan = await this.createPlan(task);
      const results = await this.executePlan(plan);
      const aggregatedResult = await this.aggregateResults(results);

      this.updateState({ status: "completed", progress: 100, completedAt: new Date().toISOString() });

      return {
        taskId: task.id,
        agentId: this.state.id,
        success: true,
        output: aggregatedResult,
        duration: Date.now() - startTime,
      };
    } catch (error: any) {
      this.updateState({ status: "failed", error: error.message });
      return {
        taskId: task.id,
        agentId: this.state.id,
        success: false,
        error: error.message,
        duration: Date.now() - startTime,
      };
    }
  }

  private async createPlan(task: AgentTask): Promise<ExecutionPlan> {
    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: this.config.systemPrompt },
        {
          role: "user",
          content: `Create an execution plan for this task:
${JSON.stringify(task, null, 2)}

Return a JSON plan with:
{
  "analysis": "Brief analysis of the request",
  "steps": [
    {
      "id": "step_1",
      "agent": "AgentName",
      "action": "what to do",
      "input": {},
      "dependencies": [],
      "priority": "high|medium|low"
    }
  ],
  "parallelGroups": [["step_1", "step_2"], ["step_3"]],
  "estimatedDuration": "time estimate",
  "fallbackStrategy": "what to do if steps fail"
}`,
        },
      ],
      temperature: this.config.temperature,
    });

    const content = response.choices[0].message.content || "{}";
    const jsonMatch = content.match(/\{[\s\S]*\}/);
    
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]) as ExecutionPlan;
    }

    return {
      analysis: "Direct execution",
      steps: [{
        id: "step_1",
        agent: "OrchestratorAgent",
        action: task.description,
        input: task.input,
        dependencies: [],
        priority: "high",
      }],
      parallelGroups: [["step_1"]],
      estimatedDuration: "unknown",
      fallbackStrategy: "retry with alternative approach",
    };
  }

  private async executePlan(plan: ExecutionPlan): Promise<StepResult[]> {
    const results: StepResult[] = [];
    const completedSteps = new Set<string>();

    for (const group of plan.parallelGroups) {
      const groupSteps = plan.steps.filter(s => group.includes(s.id));
      
      const groupPromises = groupSteps.map(async (step) => {
        const canExecute = step.dependencies.every(d => completedSteps.has(d));
        if (!canExecute) {
          return { stepId: step.id, success: false, error: "Dependencies not met" };
        }

        const agent = AGENT_REGISTRY.get(step.agent);
        if (agent) {
          const result = await agent.execute({
            id: step.id,
            type: step.action,
            description: step.action,
            input: step.input,
            priority: step.priority as any,
          });
          completedSteps.add(step.id);
          return { stepId: step.id, success: result.success, output: result.output, error: result.error };
        }
        
        const directResult = await this.executeDirectly(step);
        completedSteps.add(step.id);
        return directResult;
      });

      const groupResults = await Promise.all(groupPromises);
      results.push(...groupResults);
      
      this.updateState({ progress: Math.round((completedSteps.size / plan.steps.length) * 100) });
    }

    return results;
  }

  private async executeDirectly(step: PlanStep): Promise<StepResult> {
    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: "Execute this task directly and provide a detailed result." },
        { role: "user", content: `Task: ${step.action}\nInput: ${JSON.stringify(step.input)}` },
      ],
      temperature: 0.2,
    });

    return {
      stepId: step.id,
      success: true,
      output: response.choices[0].message.content,
    };
  }

  private async aggregateResults(results: StepResult[]): Promise<any> {
    const successfulResults = results.filter(r => r.success);
    const failedResults = results.filter(r => !r.success);

    if (failedResults.length > 0 && successfulResults.length === 0) {
      throw new Error(`All steps failed: ${failedResults.map(r => r.error).join(", ")}`);
    }

    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        {
          role: "system",
          content: "Aggregate and synthesize the results from multiple agent executions into a coherent final output.",
        },
        {
          role: "user",
          content: `Aggregate these results:
${JSON.stringify(results, null, 2)}

Provide a unified response that combines all successful outputs.`,
        },
      ],
      temperature: 0.2,
    });

    return {
      aggregatedOutput: response.choices[0].message.content,
      stepResults: results,
      summary: {
        total: results.length,
        successful: successfulResults.length,
        failed: failedResults.length,
      },
    };
  }

  getCapabilities(): AgentCapability[] {
    return [
      {
        name: "plan_execution",
        description: "Create and execute multi-step plans",
        inputSchema: z.object({ task: z.string(), context: z.record(z.any()).optional() }),
        outputSchema: z.object({ plan: z.any(), results: z.array(z.any()) }),
      },
      {
        name: "delegate_task",
        description: "Delegate tasks to specialized agents",
        inputSchema: z.object({ agent: z.string(), task: z.any() }),
        outputSchema: z.object({ result: z.any() }),
      },
      {
        name: "coordinate_workflow",
        description: "Coordinate complex multi-agent workflows",
        inputSchema: z.object({ workflow: z.any() }),
        outputSchema: z.object({ results: z.array(z.any()) }),
      },
    ];
  }
}

interface PlanStep {
  id: string;
  agent: string;
  action: string;
  input: Record<string, any>;
  dependencies: string[];
  priority: string;
}

interface ExecutionPlan {
  analysis: string;
  steps: PlanStep[];
  parallelGroups: string[][];
  estimatedDuration: string;
  fallbackStrategy: string;
}

interface StepResult {
  stepId: string;
  success: boolean;
  output?: any;
  error?: string;
}

export const orchestratorAgent = new OrchestratorAgent();

File: python_agent_tools/src/agents/orchestrator_agent.py
"""Orchestrator Agent - Coordinates and delegates tasks to specialized agents."""

from typing import Any, Dict, List, Optional
from .base_agent import BaseAgent, AgentConfig, AgentResult, AgentState
import structlog


class OrchestratorAgentConfig(AgentConfig):
    """Configuration for the Orchestrator Agent."""
    max_delegations: int = 10
    available_agents: List[str] = []
    parallel_execution: bool = True
    retry_failed: bool = True


class OrchestratorAgent(BaseAgent):
    """Super Agent that coordinates other agents and manages workflows."""
    
    name = "orchestrator"
    
    def __init__(
        self,
        config: Optional[OrchestratorAgentConfig] = None,
        tools: Optional[List] = None,
        memory = None,
    ):
        super().__init__(tools=tools, memory=memory)
        self.config = config or OrchestratorAgentConfig(name="orchestrator")
        self._registered_agents: Dict[str, BaseAgent] = {}
        self._delegation_count = 0
    
    @property
    def description(self) -> str:
        return "Coordinates and delegates tasks to specialized agents, manages complex workflows"
    
    @property
    def category(self) -> str:
        return "orchestration"
    
    @property
    def tools_used(self) -> List[str]:
        return ["orchestrate", "plan", "reason", "message"]
    
    def get_system_prompt(self) -> str:
        return """You are the Orchestrator Agent, the central coordinator of the agent system.
Your role is to:
1. Analyze incoming tasks and break them into logical subtasks
2. Delegate subtasks to the most appropriate specialized agents
3. Coordinate parallel execution when tasks are independent
4. Synthesize results from multiple agents into coherent outputs
5. Handle errors gracefully and retry failed operations
6. Track progress and provide status updates

Available specialized agents:
- ResearchAgent: Web search, information gathering, data synthesis
- CodeAgent: Code generation, review, debugging, execution
- DataAgent: Data analysis, transformation, visualization
- ContentAgent: Text generation, editing, summarization
- CommunicationAgent: Messaging, notifications, email
- BrowserAgent: Web navigation, scraping, automation
- DocumentAgent: Document creation, parsing, conversion
- QAAgent: Testing, validation, quality assurance
- SecurityAgent: Security scanning, input validation, secrets

When delegating tasks:
- Match task requirements to agent capabilities
- Provide clear context and requirements
- Set appropriate timeouts and retry policies
- Aggregate and validate results before returning"""
    
    def register_agent(self, agent: BaseAgent) -> None:
        """Register a specialized agent for delegation."""
        self._registered_agents[agent.name] = agent
        self.logger.info("agent_registered", agent_name=agent.name)
    
    def get_registered_agents(self) -> List[str]:
        """Get list of registered agent names."""
        return list(self._registered_agents.keys())
    
    async def delegate(self, agent_name: str, task: str, context: Optional[Dict[str, Any]] = None) -> AgentResult:
        """Delegate a task to a specific agent."""
        if agent_name not in self._registered_agents:
            return AgentResult(
                success=False,
                error=f"Agent '{agent_name}' not registered"
            )
        
        if self._delegation_count >= self.config.max_delegations:
            return AgentResult(
                success=False,
                error="Maximum delegation limit reached"
            )
        
        self._delegation_count += 1
        agent = self._registered_agents[agent_name]
        
        try:
            result = await agent.execute(task, context)
            return result
        except Exception as e:
            self.logger.error("delegation_failed", agent=agent_name, error=str(e))
            if self.config.retry_failed:
                return await agent.execute(task, context)
            return AgentResult(success=False, error=str(e))
    
    async def run(self, task: str, context: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Execute the orchestrator's main loop."""
        self.state = AgentState.PLANNING
        context = context or {}
        
        plan = await self.plan(task, context)
        
        self.state = AgentState.EXECUTING
        results = []
        
        for step in plan:
            result = await self.execute_step(step, context)
            results.append(result)
        
        self.state = AgentState.COMPLETED
        return {"plan": plan, "results": results}
    
    async def plan(self, task: str, context: Dict[str, Any]) -> List[str]:
        """Generate an execution plan for the task."""
        return [f"Execute: {task}"]
    
    async def execute_step(self, step: str, context: Dict[str, Any]) -> Dict[str, Any]:
        """Execute a single step of the plan."""
        self.logger.info("executing_step", step=step[:100])
        return {"step": step, "status": "completed"}
    
    async def execute(self, task: str, context: Optional[Dict[str, Any]] = None) -> AgentResult:
        """Execute a task by coordinating specialized agents."""
        self.logger.info("orchestrator_execute", task=task[:100] if task else "")
        self.state = AgentState.EXECUTING
        context = context or {}
        
        try:
            result = await self.run(task, context)
            self.state = AgentState.COMPLETED
            return AgentResult(
                success=True,
                data=result,
                metadata={"delegations": self._delegation_count}
            )
        except Exception as e:
            self.state = AgentState.ERROR
            self.logger.error("orchestrator_error", error=str(e))
            return AgentResult(success=False, error=str(e))
    
    async def initialize(self) -> None:
        """Initialize the orchestrator agent."""
        await super().initialize()
        self._delegation_count = 0
        self.logger.info("orchestrator_initialized", registered_agents=len(self._registered_agents))
    
    async def shutdown(self) -> None:
        """Shutdown the orchestrator agent."""
        for agent in self._registered_agents.values():
            await agent.shutdown()
        await super().shutdown()
        self.logger.info("orchestrator_shutdown")
